cdap-docs/developers-manual/source/building-blocks/workflows.rst

```diff
@@ -86,12 +94,114 @@ implement the ``run()`` method::
 The custom action then can be added to the workflow using the ``addAction()`` method as
 shown above.
 
+.. _workflow-concurrent:
+
 Concurrent Workflows
 --------------------
 By default, a workflow runs sequentially. Multiple instances of a workflow can be run
 concurrently. To enable concurrent runs for a workflow, set its runtime argument
 ``concurrent.runs.enabled`` to ``true``.
 
+
+.. _workflow_token:
+
+Workflow Tokens
+===============
+
+In addition to passing the control flow from one node to the next, a **workflow token** is
+passed, available to each of the programs in the workflow. This allows programs to:
+
+- pass custom data (such as metrics, a status, or an error code) from one program in the 
+  workflow to the next subsequent programs; 
+- query and set the data in the token;
+- fetch the data from the token which was set by a specific node; and
+- determine the name of the node which most recently set the token value for a specific key;
+  e.g., the node who last set the ERROR flag in the token.
+  
+The last example is intended to allow appropriate action to be taken in response, such as
+logging, modifying the conditional execution, or terminating the execution of workflow.
+
+Execution in the workflow can be made conditional, based on the information contained in
+the token. Execution can be terminated if a node in the workflow produces unexpected
+results.
+
+Once a run is completed, you can query the tokens from past workflow runs for analyses that
+determine which node was executed more frequently and when. You can retrieve the token values
+that were added by a specific node in the workflow to debug the flow of execution.
+
+Scope
+-----
+Two scopes |---| *System* and *User* |---| are provided for workflow keys. CDAP adds keys
+(such as MapReduce counters) under the *System* scope. User programs add their keys under
+the *User* scope.
+
+Putting and Getting Token Values
+--------------------------------
+When a value is put into a token, it is stored under a specific key. Both keys and their
+corresponding values must be non-null. The token stores additional information about the 
+context in which the key is being set, such as the unique name of the workflow node. 
+
+To put a value to a token, first obtain access to the token from the workflow context, and
+then set a value for a specific key, as shown in this example from a MapReduce mapper::
+
+  public static class MyVerifier extends Mapper<LongWritable, Text, Text, NullWritable> {
+    public void map(LongWritable key, Text value, Context context)
+      throws IOException, InterruptedException {
+      String profile;
+      WorkflowToken workflowToken = context.getWorkflowToken();
+      if (workflowToken != null) {
```

Comment:
Should we point out somewhere that this null check is only required because this Mapper could be used outside of a workflow. When run from within a workflow, it is guaranteed to be non-null.